// neutrinos-seed-services/esm5/src/app/neutrinos-seed-services/n-login.service.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
import { Injectable, EventEmitter, Output } from '@angular/core';
import { HttpClient } from '@angular/common/http';
import { NSystemService } from './n-system.service';
import { NTokenService } from './n-token.service';
import { NPubSubService } from './n-pubSub.service';
import { NSessionStorageService } from './n-sessionStorage.service';
import { NLocalStorageService } from './n-localStorage.service';
import { NNotificationService } from './n-notification.service';
import 'rxjs/add/operator/map';
var NLoginService = /** @class */ (function () {
    function NLoginService(http, pubSubService, notificationService, nLocalStorageService, nTokenService) {
        this.http = http;
        this.pubSubService = pubSubService;
        this.notificationService = notificationService;
        this.nLocalStorageService = nLocalStorageService;
        this.nTokenService = nTokenService;
        this.loginCompleted = new EventEmitter();
        this.systemService = NSystemService.getInstance();
        // this.nTokenService = new NTokenService();
        this.nSessionStorage = new NSessionStorageService();
        // this.nLocalStorageService = new NLocalStorageService();
    }
    /**
     * @param {?} userName
     * @param {?} password
     * @param {?=} isRemember
     * @return {?}
     */
    NLoginService.prototype.login = /**
     * @param {?} userName
     * @param {?} password
     * @param {?=} isRemember
     * @return {?}
     */
    function (userName, password, isRemember) {
        var _this = this;
        this.appProperties = this.systemService.getVal('properties');
        this.loginUrl = this.systemService.getAuthUrl() + this.appProperties.appName;
        this.uuid = this.nLocalStorageService.getValue('uuid');
        if (!this.uuid) {
            this.uuid = this.nLocalStorageService.checkDeviceId();
        }
        this.details = {
            username: userName,
            password: password,
        };
        this.details.platformDetails = this.systemService.getPlatformDetails(this.systemService.checkDevice());
        this.details.platformDetails['uuid'] = this.uuid;
        return this.http.post(this.loginUrl, JSON.stringify(this.details)).map(function (result) {
            var /** @type {?} */ tokensObj = result;
            if (tokensObj) {
                _this.nTokenService.updateTokens(tokensObj, isRemember);
            }
            // TODO chris array of supported pushes currently only support APNS and Firebase
            if ((_this.systemService.getVal('firebaseSenderId') != 'FIREBASE_SENDER_ID' && _this.systemService.getVal('firebaseAuthKey') != 'FIREBASE_AUTH_KEY')
                || (_this.systemService.getVal('pushType') === 'APNS' && _this.systemService.isIOS())) {
                _this.pubSubService.$pub('firebaseRegister');
            }
            _this.pubSubService.$pub('loginComplete');
            return (result);
        }, function (error) {
            return (error);
        });
    };
    /**
     * @return {?}
     */
    NLoginService.prototype.isLoggedIn = /**
     * @return {?}
     */
    function () {
        var _this = this;
        return this.nLocalStorageService.initStorage().then(function (result) {
            if (_this.nSessionStorage.getValue('accessToken') && _this.nSessionStorage.getValue('refreshToken') &&
                _this.nSessionStorage.getValue('accessToken') != 'null' && _this.nSessionStorage.getValue('refreshToken') != 'null') {
                return true;
            }
            return false;
        }).catch(function (error) {
            return false;
        });
    };
    NLoginService.decorators = [
        { type: Injectable },
    ];
    /** @nocollapse */
    NLoginService.ctorParameters = function () { return [
        { type: HttpClient, },
        { type: NPubSubService, },
        { type: NNotificationService, },
        { type: NLocalStorageService, },
        { type: NTokenService, },
    ]; };
    NLoginService.propDecorators = {
        "loginCompleted": [{ type: Output },],
    };
    return NLoginService;
}());
export { NLoginService };
function NLoginService_tsickle_Closure_declarations() {
    /** @type {!Array<{type: !Function, args: (undefined|!Array<?>)}>} */
    NLoginService.decorators;
    /**
     * @nocollapse
     * @type {function(): !Array<(null|{type: ?, decorators: (undefined|!Array<{type: !Function, args: (undefined|!Array<?>)}>)})>}
     */
    NLoginService.ctorParameters;
    /** @type {!Object<string,!Array<{type: !Function, args: (undefined|!Array<?>)}>>} */
    NLoginService.propDecorators;
    /** @type {?} */
    NLoginService.prototype.loginUrl;
    /** @type {?} */
    NLoginService.prototype.appProperties;
    /** @type {?} */
    NLoginService.prototype.systemService;
    /** @type {?} */
    NLoginService.prototype.nSessionStorage;
    /** @type {?} */
    NLoginService.prototype.uuid;
    /** @type {?} */
    NLoginService.prototype.details;
    /** @type {?} */
    NLoginService.prototype.loginCompleted;
    /** @type {?} */
    NLoginService.prototype.http;
    /** @type {?} */
    NLoginService.prototype.pubSubService;
    /** @type {?} */
    NLoginService.prototype.notificationService;
    /** @type {?} */
    NLoginService.prototype.nLocalStorageService;
    /** @type {?} */
    NLoginService.prototype.nTokenService;
}
