// neutrinos-seed-services/esm5/src/app/neutrinos-seed-services/n-sessionStorage.service.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
import { Injectable } from '@angular/core';
var NSessionStorageService = /** @class */ (function () {
    function NSessionStorageService() {
        NSessionStorageService.sessionStorageCache = sessionStorage;
    }
    /**
     * @return {?}
     */
    NSessionStorageService.prototype.getSessionStorage = /**
     * @return {?}
     */
    function () {
        return NSessionStorageService.sessionStorageCache;
    };
    /**
     * @param {?} key
     * @param {?} value
     * @return {?}
     */
    NSessionStorageService.prototype.setValue = /**
     * @param {?} key
     * @param {?} value
     * @return {?}
     */
    function (key, value) {
        NSessionStorageService.sessionStorageCache[key] = value;
        sessionStorage.setItem(key, value);
    };
    /**
     * @param {?} key
     * @return {?}
     */
    NSessionStorageService.prototype.getValue = /**
     * @param {?} key
     * @return {?}
     */
    function (key) {
        if (!NSessionStorageService.sessionStorageCache[key]) {
            return null;
        }
        try {
            var /** @type {?} */ obj = NSessionStorageService.sessionStorageCache[key];
            return JSON.parse(obj);
        }
        catch (/** @type {?} */ error) {
            return NSessionStorageService.sessionStorageCache[key];
        }
    };
    /**
     * @param {?} key
     * @return {?}
     */
    NSessionStorageService.prototype.remove = /**
     * @param {?} key
     * @return {?}
     */
    function (key) {
        if (NSessionStorageService.sessionStorageCache.hasOwnProperty(key)) {
            delete NSessionStorageService.sessionStorageCache[key];
            sessionStorage.removeItem(key);
        }
    };
    /**
     * @return {?}
     */
    NSessionStorageService.prototype.clearSessionStorage = /**
     * @return {?}
     */
    function () {
        NSessionStorageService.sessionStorageCache = {};
        sessionStorage.clear();
    };
    NSessionStorageService.sessionStorageCache = {};
    NSessionStorageService.decorators = [
        { type: Injectable },
    ];
    /** @nocollapse */
    NSessionStorageService.ctorParameters = function () { return []; };
    return NSessionStorageService;
}());
export { NSessionStorageService };
function NSessionStorageService_tsickle_Closure_declarations() {
    /** @type {!Array<{type: !Function, args: (undefined|!Array<?>)}>} */
    NSessionStorageService.decorators;
    /**
     * @nocollapse
     * @type {function(): !Array<(null|{type: ?, decorators: (undefined|!Array<{type: !Function, args: (undefined|!Array<?>)}>)})>}
     */
    NSessionStorageService.ctorParameters;
    /** @type {?} */
    NSessionStorageService.sessionStorageCache;
}
