// neutrinos-seed-services/esm5/src/app/neutrinos-seed-services/neutrinos-seed-services.module.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
import { NgModule, APP_INITIALIZER } from '@angular/core';
import { CommonModule } from '@angular/common';
import { HTTP_INTERCEPTORS } from '@angular/common/http';
import { NAuthGuardService } from './n-authGaurd.service';
import { NDataModelService } from './n-dataModel.service';
import { NHttpService } from './n-HTTP.service';
import { NHTTPLoaderService } from './n-HTTPLoader.service';
import { NLocalStorageService } from './n-localStorage.service';
import { NLoginService } from './n-login.service';
import { NLogoutService } from './n-logout.service';
import { NNotificationService } from './n-notification.service';
import { NSnackbarService } from './n-snackbar.service';
import { NPubSubService } from './n-pubSub.service';
import { NSessionStorageService } from './n-sessionStorage.service';
import { NTokenService } from './n-token.service';
import { HttpClientModule } from '@angular/common/http';
import { RouterModule } from '@angular/router';
import { HttpModule } from '@angular/http';
import { NgForageModule } from 'ngforage';
import { SnackComponent } from './snackbar/snack/snack.component';
import { SnackbarComponent } from './snackbar/snackbar/snackbar.component';
/**
 * @param {?} startupService
 * @return {?}
 */
export function startupServiceFactory(startupService) {
    var /** @type {?} */ res = function () {
        // console.log('module before initstorage call');
        return startupService.initStorage();
    };
    return res;
}
var NeutrinosSeedServicesModule = /** @class */ (function () {
    function NeutrinosSeedServicesModule() {
    }
    NeutrinosSeedServicesModule.decorators = [
        { type: NgModule, args: [{
                    imports: [
                        CommonModule,
                        HttpClientModule,
                        RouterModule,
                        HttpModule,
                        NgForageModule.forRoot()
                    ],
                    providers: [
                        {
                            provide: HTTP_INTERCEPTORS,
                            useClass: NHttpService,
                            multi: true
                        },
                        {
                            // Provider for APP_INITIALIZER
                            provide: APP_INITIALIZER,
                            useFactory: startupServiceFactory,
                            deps: [NLocalStorageService],
                            multi: true
                        },
                        NAuthGuardService,
                        NDataModelService,
                        NHttpService,
                        NHTTPLoaderService,
                        NLocalStorageService,
                        NLoginService,
                        NLogoutService,
                        NNotificationService,
                        NSnackbarService,
                        NPubSubService,
                        NSessionStorageService,
                        NTokenService
                    ],
                    declarations: [
                        SnackComponent,
                        SnackbarComponent
                    ],
                    exports: [
                        SnackComponent,
                        SnackbarComponent
                    ]
                },] },
    ];
    return NeutrinosSeedServicesModule;
}());
export { NeutrinosSeedServicesModule };
function NeutrinosSeedServicesModule_tsickle_Closure_declarations() {
    /** @type {!Array<{type: !Function, args: (undefined|!Array<?>)}>} */
    NeutrinosSeedServicesModule.decorators;
    /**
     * @nocollapse
     * @type {function(): !Array<(null|{type: ?, decorators: (undefined|!Array<{type: !Function, args: (undefined|!Array<?>)}>)})>}
     */
    NeutrinosSeedServicesModule.ctorParameters;
}
