// neutrinos-seed-services/esm2015/src/app/neutrinos-seed-services/n-HTTPLoader.service.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
import { Injectable } from '@angular/core';
import { Subject } from 'rxjs/Subject';
import { NLogoutService } from './n-logout.service';
import { NPubSubService } from './n-pubSub.service';
export class NHTTPLoaderService {
    /**
     * @param {?} nLogoutService
     * @param {?} nPubSubService
     */
    constructor(nLogoutService, nPubSubService) {
        this.nLogoutService = nLogoutService;
        this.nPubSubService = nPubSubService;
        this.pendingRequests = 0;
        this.showLoading = false;
        this._isHTTPRequestInProgress = new Subject();
        this._isHTTPRequestInProgress$ = this._isHTTPRequestInProgress.asObservable();
    }
    /**
     * @param {?} bool
     * @return {?}
     */
    isHTTPRequestInProgress(bool) {
        if (bool) {
            this.pendingRequests++;
            if (!this.showLoading) {
                this._isHTTPRequestInProgress.next(bool);
            }
            this.showLoading = true;
        }
        else {
            this.pendingRequests--;
            if (this.pendingRequests <= 0) {
                if (this.showLoading) {
                    this._isHTTPRequestInProgress.next(bool);
                }
                this.showLoading = false;
            }
        }
    }
    /**
     * @param {?} error
     * @return {?}
     */
    alertError(error) {
        if (error.status < 200 || error.status > 500) {
            if (error.status === 0) {
                this.nPubSubService.$pub('openSnackBar', 'Connectivity issue');
            }
            else {
                this.nPubSubService.$pub('openSnackBar', 'Response failure');
            }
        }
        else if (error.status === 401) {
            this.nLogoutService.logout();
            if (error.error) {
                this.nPubSubService.$pub('openSnackBar', error.error);
            }
            else {
                this.nPubSubService.$pub('openSnackBar', error);
            }
        }
        else if (error.error) {
            if (error.error instanceof Object) {
                this.nPubSubService.$pub('openSnackBar', error.error.message);
            }
            else {
                this.nPubSubService.$pub('openSnackBar', error.error);
            }
        }
        else {
            this.nPubSubService.$pub('openSnackBar', 'Response failure');
        }
    }
}
NHTTPLoaderService.decorators = [
    { type: Injectable },
];
/** @nocollapse */
NHTTPLoaderService.ctorParameters = () => [
    { type: NLogoutService, },
    { type: NPubSubService, },
];
function NHTTPLoaderService_tsickle_Closure_declarations() {
    /** @type {!Array<{type: !Function, args: (undefined|!Array<?>)}>} */
    NHTTPLoaderService.decorators;
    /**
     * @nocollapse
     * @type {function(): !Array<(null|{type: ?, decorators: (undefined|!Array<{type: !Function, args: (undefined|!Array<?>)}>)})>}
     */
    NHTTPLoaderService.ctorParameters;
    /** @type {?} */
    NHTTPLoaderService.prototype.pendingRequests;
    /** @type {?} */
    NHTTPLoaderService.prototype.showLoading;
    /** @type {?} */
    NHTTPLoaderService.prototype._isHTTPRequestInProgress;
    /** @type {?} */
    NHTTPLoaderService.prototype._isHTTPRequestInProgress$;
    /** @type {?} */
    NHTTPLoaderService.prototype.nLogoutService;
    /** @type {?} */
    NHTTPLoaderService.prototype.nPubSubService;
}
