// neutrinos-seed-services/esm2015/src/app/neutrinos-seed-services/n-login.service.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
import { Injectable, EventEmitter, Output } from '@angular/core';
import { HttpClient } from '@angular/common/http';
import { NSystemService } from './n-system.service';
import { NTokenService } from './n-token.service';
import { NPubSubService } from './n-pubSub.service';
import { NSessionStorageService } from './n-sessionStorage.service';
import { NLocalStorageService } from './n-localStorage.service';
import { NNotificationService } from './n-notification.service';
import 'rxjs/add/operator/map';
export class NLoginService {
    /**
     * @param {?} http
     * @param {?} pubSubService
     * @param {?} notificationService
     * @param {?} nLocalStorageService
     * @param {?} nTokenService
     */
    constructor(http, pubSubService, notificationService, nLocalStorageService, nTokenService) {
        this.http = http;
        this.pubSubService = pubSubService;
        this.notificationService = notificationService;
        this.nLocalStorageService = nLocalStorageService;
        this.nTokenService = nTokenService;
        this.loginCompleted = new EventEmitter();
        this.systemService = NSystemService.getInstance();
        // this.nTokenService = new NTokenService();
        this.nSessionStorage = new NSessionStorageService();
        // this.nLocalStorageService = new NLocalStorageService();
    }
    /**
     * @param {?} userName
     * @param {?} password
     * @param {?=} isRemember
     * @return {?}
     */
    login(userName, password, isRemember) {
        this.appProperties = this.systemService.getVal('properties');
        this.loginUrl = this.systemService.getAuthUrl() + this.appProperties.appName;
        this.uuid = this.nLocalStorageService.getValue('uuid');
        if (!this.uuid) {
            this.uuid = this.nLocalStorageService.checkDeviceId();
        }
        this.details = {
            username: userName,
            password: password,
        };
        this.details.platformDetails = this.systemService.getPlatformDetails(this.systemService.checkDevice());
        this.details.platformDetails['uuid'] = this.uuid;
        return this.http.post(this.loginUrl, JSON.stringify(this.details)).map(result => {
            const /** @type {?} */ tokensObj = result;
            if (tokensObj) {
                this.nTokenService.updateTokens(tokensObj, isRemember);
            }
            // TODO chris array of supported pushes currently only support APNS and Firebase
            if ((this.systemService.getVal('firebaseSenderId') != 'FIREBASE_SENDER_ID' && this.systemService.getVal('firebaseAuthKey') != 'FIREBASE_AUTH_KEY')
                || (this.systemService.getVal('pushType') === 'APNS' && this.systemService.isIOS())) {
                this.pubSubService.$pub('firebaseRegister');
            }
            this.pubSubService.$pub('loginComplete');
            return (result);
        }, error => {
            return (error);
        });
    }
    /**
     * @return {?}
     */
    isLoggedIn() {
        return this.nLocalStorageService.initStorage().then(result => {
            if (this.nSessionStorage.getValue('accessToken') && this.nSessionStorage.getValue('refreshToken') &&
                this.nSessionStorage.getValue('accessToken') != 'null' && this.nSessionStorage.getValue('refreshToken') != 'null') {
                return true;
            }
            return false;
        }).catch(error => {
            return false;
        });
    }
}
NLoginService.decorators = [
    { type: Injectable },
];
/** @nocollapse */
NLoginService.ctorParameters = () => [
    { type: HttpClient, },
    { type: NPubSubService, },
    { type: NNotificationService, },
    { type: NLocalStorageService, },
    { type: NTokenService, },
];
NLoginService.propDecorators = {
    "loginCompleted": [{ type: Output },],
};
function NLoginService_tsickle_Closure_declarations() {
    /** @type {!Array<{type: !Function, args: (undefined|!Array<?>)}>} */
    NLoginService.decorators;
    /**
     * @nocollapse
     * @type {function(): !Array<(null|{type: ?, decorators: (undefined|!Array<{type: !Function, args: (undefined|!Array<?>)}>)})>}
     */
    NLoginService.ctorParameters;
    /** @type {!Object<string,!Array<{type: !Function, args: (undefined|!Array<?>)}>>} */
    NLoginService.propDecorators;
    /** @type {?} */
    NLoginService.prototype.loginUrl;
    /** @type {?} */
    NLoginService.prototype.appProperties;
    /** @type {?} */
    NLoginService.prototype.systemService;
    /** @type {?} */
    NLoginService.prototype.nSessionStorage;
    /** @type {?} */
    NLoginService.prototype.uuid;
    /** @type {?} */
    NLoginService.prototype.details;
    /** @type {?} */
    NLoginService.prototype.loginCompleted;
    /** @type {?} */
    NLoginService.prototype.http;
    /** @type {?} */
    NLoginService.prototype.pubSubService;
    /** @type {?} */
    NLoginService.prototype.notificationService;
    /** @type {?} */
    NLoginService.prototype.nLocalStorageService;
    /** @type {?} */
    NLoginService.prototype.nTokenService;
}
