// neutrinos-seed-services/esm2015/src/app/neutrinos-seed-services/n-pubSub.service.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
import { Injectable } from '@angular/core';
import { Subject } from 'rxjs/Subject';
const /** @type {?} */ ServiceName = "PubSub Service";
export class NPubSubService {
    constructor() {
        this.events = {};
    }
    /**
     * @param {?} event
     * @param {?=} callback
     * @param {?=} error
     * @param {?=} complete
     * @return {?}
     */
    $sub(event, callback, error, complete) {
        if (!event) {
            throw new Error(`[${ServiceName}] => Subscription method must get event name.`);
        }
        if (this.events[event] === undefined) {
            this.events[event] = new Subject();
        }
        if (typeof callback !== 'function') {
            return this.events[event].asObservable();
        }
        else {
            return this.events[event].asObservable().subscribe(callback, error, complete);
        }
    }
    /**
     * @param {?} event
     * @param {?=} eventObject
     * @return {?}
     */
    $pub(event, eventObject) {
        if (!event) {
            throw new Error(`[${ServiceName}] => Publish method must get event name.`);
        }
        else if (!this.events[event]) {
            return;
        }
        this.events[event].next(eventObject);
    }
}
NPubSubService.decorators = [
    { type: Injectable },
];
/** @nocollapse */
NPubSubService.ctorParameters = () => [];
function NPubSubService_tsickle_Closure_declarations() {
    /** @type {!Array<{type: !Function, args: (undefined|!Array<?>)}>} */
    NPubSubService.decorators;
    /**
     * @nocollapse
     * @type {function(): !Array<(null|{type: ?, decorators: (undefined|!Array<{type: !Function, args: (undefined|!Array<?>)}>)})>}
     */
    NPubSubService.ctorParameters;
    /** @type {?} */
    NPubSubService.prototype.events;
}
/**
 * @record
 */
export function IPubSubService() { }
function IPubSubService_tsickle_Closure_declarations() {
    /** @type {?} */
    IPubSubService.prototype.$pub;
    /** @type {?} */
    IPubSubService.prototype.$sub;
    /** @type {?} */
    IPubSubService.prototype.$sub;
    /** @type {?} */
    IPubSubService.prototype.$sub;
    /** @type {?} */
    IPubSubService.prototype.$sub;
}
/**
 * @record
 */
function I$sub() { }
function I$sub_tsickle_Closure_declarations() {
    /* TODO: handle strange member:
    (event: string): Observable<any>;
    */
    /* TODO: handle strange member:
    (event: string, callback: (value: any) => void): Subscription;
    */
    /* TODO: handle strange member:
    (event: string, callback: (value: any) => void, error: (error: any) => void): Subscription;
    */
    /* TODO: handle strange member:
    (event: string, callback: (value: any) => void, error: (error: any) => void, complete: () => void): Subscription;
    */
}
