// neutrinos-seed-services/esm2015/src/app/neutrinos-seed-services/n-snackbar.service.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
import { Injectable } from '@angular/core';
import { Subject } from 'rxjs/Subject';
import { NSystemService } from './n-system.service';
import { NPubSubService } from './n-pubSub.service';
export class NSnackbarService {
    /**
     * @param {?} nPubSubService
     */
    constructor(nPubSubService) {
        this.nPubSubService = nPubSubService;
        this.snackService = new Subject();
        if (NSystemService.getInstance().getVal('useDefaultExceptionUI')) {
            this.snackBarListner = this.nPubSubService.$sub('openSnackBar', (msg) => {
                this.openSnackBar(msg);
            });
        }
    }
    /**
     * @return {?}
     */
    get() {
        return this.snackService.asObservable();
    }
    /**
     * @param {?} data
     * @return {?}
     */
    add(data) {
        this.snackService.next({
            action: 'add',
            data: data
        });
    }
    /**
     * @param {?} id
     * @return {?}
     */
    remove(id) {
        this.snackService.next({ action: 'remove', id: id });
    }
    /**
     * @return {?}
     */
    clear() {
        this.snackService.next({ action: 'clear' });
    }
    /**
     * @param {?} message
     * @param {?=} timeout
     * @return {?}
     */
    openSnackBar(message, timeout) {
        this.add({
            msg: message,
            timeout: timeout || 5000,
            action: {
                text: 'Close',
                onClick: (snack) => {
                },
            },
            onAdd: (snack) => {
            },
            onRemove: (snack) => {
            }
        });
    }
}
NSnackbarService.decorators = [
    { type: Injectable },
];
/** @nocollapse */
NSnackbarService.ctorParameters = () => [
    { type: NPubSubService, },
];
function NSnackbarService_tsickle_Closure_declarations() {
    /** @type {!Array<{type: !Function, args: (undefined|!Array<?>)}>} */
    NSnackbarService.decorators;
    /**
     * @nocollapse
     * @type {function(): !Array<(null|{type: ?, decorators: (undefined|!Array<{type: !Function, args: (undefined|!Array<?>)}>)})>}
     */
    NSnackbarService.ctorParameters;
    /** @type {?} */
    NSnackbarService.prototype.snackBarListner;
    /** @type {?} */
    NSnackbarService.prototype.snackService;
    /** @type {?} */
    NSnackbarService.prototype.nPubSubService;
}
