// neutrinos-seed-services/esm2015/src/app/neutrinos-seed-services/n-system.service.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
import * as platform from 'platform';
export class NSystemService {
    /**
     * @return {?}
     */
    static getInstance() {
        if (!this.instance) {
            this.instance = new NSystemService();
        }
        return this.instance;
    }
    constructor() {
        this._deviceType = this.checkDevice();
    }
    /**
     * Returna 'mobile' or 'browser' based on the platform
     * @return {?}
     */
    checkDevice() {
        if (window['device'] && (window['device']['platform'] != 'browser' || window['device']['platform'] != 'Browser')) {
            return 'mobile';
        }
        else if (window['device'] && (window['device']['platform'] == 'browser' || window['device']['platform'] == 'Browser')) {
            return 'cordova_browser';
        }
        else {
            return 'browser';
        }
    }
    /**
     * @return {?}
     */
    get deviceType() {
        return this._deviceType;
    }
    /**
     * @param {?} key
     * @return {?}
     */
    getVal(key) {
        if (!NSystemService.envProperties) {
            NSystemService.envProperties = window['neutrinos']['environments'];
            this.properties = NSystemService.envProperties.properties;
        }
        if (key == 'properties') {
            if (NSystemService.envProperties) {
                return NSystemService.envProperties.properties;
            }
            else {
                return undefined;
            }
        }
        else {
            if (NSystemService.envProperties && NSystemService.envProperties.properties) {
                return NSystemService.envProperties.properties[key];
            }
            else {
                return undefined;
            }
        }
    }
    /**
     * @return {?}
     */
    isAndroid() {
        if (window['device'] && window['device']['platform'] == 'Android') {
            return true;
        }
        else {
            return false;
        }
    }
    /**
     * @return {?}
     */
    isIOS() {
        if (window['device'] && window['device']['platform'] == 'iOS') {
            return true;
        }
        else {
            return false;
        }
    }
    /**
     * @return {?}
     */
    getAndroidVersion() {
        if (this.isAndroid()) {
            return window['device']['version'];
        }
        else {
            return '';
        }
    }
    /**
     * @return {?}
     */
    getAuthUrl() {
        // this.properties = this.getVal('properties');
        return this.properties.baseUrl + this.properties.tenantName + '/auth/';
    }
    /**
     * @return {?}
     */
    getTenantUrl() {
        return this.properties.baseUrl + this.properties.tenantName + '/';
    }
    /**
     * @return {?}
     */
    getAppUrl() {
        return this.properties.baseUrl + this.properties.tenantName + '/proxy/' + this.properties.appName + '/';
    }
    /**
     * @return {?}
     */
    getProxyUrl() {
        return this.properties.baseUrl + this.properties.tenantName + '/proxy/';
    }
    /**
     * @return {?}
     */
    getDataModelUrl() {
        return this.properties.baseUrl + this.properties.tenantName + '/datamodel/' + this.properties.appDataSource + '/' + this.properties.appName + '/';
    }
    /**
     * @return {?}
     */
    getFileIOUrl() {
        return this.properties.baseUrl + this.properties.tenantName + '/fileio/' +
            this.properties.appDataSource + '/' + this.properties.appName + '/';
    }
    /**
     * @param {?} type
     * @return {?}
     */
    getPlatformDetails(type) {
        const /** @type {?} */ pDetails = {
            platform: platform.os.family,
            platformVersion: platform.os.version,
        };
        if (type === 'browser') {
            return Object.assign(pDetails, {
                type: 'browser',
                browser: platform.name,
                browserVersion: platform.version,
            });
        }
        else if (type === 'mobile' || type === 'cordova_browser') {
            return Object.assign(pDetails, {
                type: 'mobile',
                model: window['device'].model,
                manufacturer: window['device'].manufacturer,
                isVirtual: window['device'].isVirtual
            });
        }
    }
}
function NSystemService_tsickle_Closure_declarations() {
    /** @type {?} */
    NSystemService.instance;
    /** @type {?} */
    NSystemService.envProperties;
    /** @type {?} */
    NSystemService.prototype._deviceType;
    /** @type {?} */
    NSystemService.prototype.uuid;
    /** @type {?} */
    NSystemService.prototype.properties;
}
