// n-sheet/esm2015/lib/n-sheet.service.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Injectable } from '@angular/core';
import { HttpClient } from '@angular/common/http';
import * as i0 from "@angular/core";
import * as i1 from "@angular/common/http";
export class NSheetService {
    /**
     * @param {?} http
     */
    constructor(http) {
        this.http = http;
    }
    /**
     * @param {?=} sheetOptions
     * @param {?=} htmlContent
     * @return {?}
     */
    getSheet(sheetOptions, htmlContent) {
        return new Promise((/**
         * @param {?} resolve
         * @param {?} reject
         * @return {?}
         */
        (resolve, reject) => {
            if (typeof sheetOptions === 'object') {
                if (sheetOptions.hasOwnProperty('url')) {
                    delete sheetOptions.retPure;
                    sheetOptions.httpMethod = (sheetOptions.hasOwnProperty('httpMethod') && (sheetOptions.httpMethod.toLowerCase().trim() === 'get' || sheetOptions.httpMethod.toLowerCase().trim() === 'put')) ? sheetOptions.httpMethod.toLowerCase().trim() : 'post';
                    /** @type {?} */
                    var body = (sheetOptions.httpMethod.toLowerCase().trim() === 'get') ? {} : sheetOptions;
                    this.executeCallback(sheetOptions).then((/**
                     * @param {?} res
                     * @return {?}
                     */
                    res => {
                        this.http[sheetOptions.httpMethod.toLowerCase().trim()](sheetOptions.url, body).subscribe((/**
                         * @param {?} response
                         * @return {?}
                         */
                        response => {
                            return resolve(response);
                        }), (/**
                         * @param {?} error
                         * @return {?}
                         */
                        error => {
                            return reject(error);
                        }));
                    })).catch((/**
                     * @param {?} err
                     * @return {?}
                     */
                    err => {
                        return reject(err);
                    }));
                }
                else {
                    return reject('No url');
                }
            }
            else if (htmlContent) {
                return resolve({ excelAsHtml: htmlContent });
            }
            else {
                return reject('Please specify correct values for sheetOptions / htmlContent');
            }
        }));
    }
    /**
     * @param {?} sheetOptions
     * @return {?}
     */
    executeCallback(sheetOptions) {
        return new Promise((/**
         * @param {?} resolve
         * @param {?} reject
         * @return {?}
         */
        (resolve, reject) => {
            if (sheetOptions.hasOwnProperty('preHttpRequestCallback') && (sheetOptions.preHttpRequestCallback instanceof Promise)) {
                sheetOptions.preHttpRequestCallback.then((/**
                 * @param {?} res
                 * @return {?}
                 */
                res => {
                    return resolve();
                })).catch((/**
                 * @param {?} err
                 * @return {?}
                 */
                err => {
                    return reject('Promise function returned a reject');
                }));
            }
            else if (sheetOptions.hasOwnProperty('preHttpRequestCallback') && (typeof sheetOptions.preHttpRequestCallback === 'boolean')) {
                if (sheetOptions.preHttpRequestCallback === true) {
                    return resolve();
                }
                else {
                    return reject('Boolean function returned a false');
                }
            }
            else if (sheetOptions.hasOwnProperty('preHttpRequestCallback') && (typeof sheetOptions.preHttpRequestCallback === 'function')) {
                if (sheetOptions.preHttpRequestCallback() === true) {
                    return resolve();
                }
                else {
                    return reject('Callback function either returned a false or did not return a boolean value');
                }
            }
            else {
                return resolve();
            }
        }));
    }
}
NSheetService.decorators = [
    { type: Injectable, args: [{
                providedIn: 'root'
            },] }
];
/** @nocollapse */
NSheetService.ctorParameters = () => [
    { type: HttpClient }
];
/** @nocollapse */ NSheetService.ngInjectableDef = i0.defineInjectable({ factory: function NSheetService_Factory() { return new NSheetService(i0.inject(i1.HttpClient)); }, token: NSheetService, providedIn: "root" });
if (false) {
    /**
     * @type {?}
     * @private
     */
    NSheetService.prototype.http;
}
