// n-sheet/esm2015/lib/n-sheet.directive.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Directive, Input, Output, EventEmitter, ElementRef } from '@angular/core';
import { NSheetService } from './n-sheet.service';
import { HttpClient } from '@angular/common/http';
export class NSheetDirective {
    /**
     * @param {?} sheetService
     * @param {?} http
     * @param {?} el
     */
    constructor(sheetService, http, el) {
        this.sheetService = sheetService;
        this.http = http;
        this.el = el;
        this.onsuccess = new EventEmitter();
        this.onerror = new EventEmitter();
    }
    /**
     * @return {?}
     */
    ngOnInit() {
        this.getSheet()
            .then((/**
         * @param {?} success
         * @return {?}
         */
        success => this.onsuccess.emit(success)))
            .catch((/**
         * @param {?} error
         * @return {?}
         */
        error => this.onerror.emit(error)));
    }
    /**
     * @return {?}
     */
    getSheet() {
        return new Promise((/**
         * @param {?} resolve
         * @param {?} reject
         * @return {?}
         */
        (resolve, reject) => {
            this.sheetService.getSheet(this.sheetOptions, this.htmlContent).then((/**
             * @param {?} res
             * @return {?}
             */
            res => {
                if (res.hasOwnProperty('excelAsHtml') && res['excelAsHtml'].length > 0 && !Array.isArray(res['excelAsHtml'])) {
                    if (!(typeof this.sheetOptions === 'object' && this.sheetOptions.hasOwnProperty('doAppend') && this.sheetOptions.doAppend === true)) {
                        this.el.nativeElement.innerHTML = "";
                    }
                    /** @type {?} */
                    const fragment = document.createRange().createContextualFragment(res['excelAsHtml']);
                    this.el.nativeElement.appendChild(fragment);
                    return resolve(res);
                }
                else {
                    return reject(res);
                }
            })).catch((/**
             * @param {?} err
             * @return {?}
             */
            err => reject(err)));
        }));
    }
}
NSheetDirective.decorators = [
    { type: Directive, args: [{
                selector: '[n-sheet]'
            },] }
];
/** @nocollapse */
NSheetDirective.ctorParameters = () => [
    { type: NSheetService },
    { type: HttpClient },
    { type: ElementRef }
];
NSheetDirective.propDecorators = {
    sheetOptions: [{ type: Input }],
    htmlContent: [{ type: Input }],
    onsuccess: [{ type: Output }],
    onerror: [{ type: Output }]
};
if (false) {
    /** @type {?} */
    NSheetDirective.prototype.sheetOptions;
    /** @type {?} */
    NSheetDirective.prototype.htmlContent;
    /** @type {?} */
    NSheetDirective.prototype.onsuccess;
    /** @type {?} */
    NSheetDirective.prototype.onerror;
    /**
     * @type {?}
     * @private
     */
    NSheetDirective.prototype.sheetService;
    /**
     * @type {?}
     * @private
     */
    NSheetDirective.prototype.http;
    /**
     * @type {?}
     * @private
     */
    NSheetDirective.prototype.el;
}
