// n-sheet/esm5/lib/n-sheet.directive.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Directive, Input, Output, EventEmitter, ElementRef } from '@angular/core';
import { NSheetService } from './n-sheet.service';
import { HttpClient } from '@angular/common/http';
var NSheetDirective = /** @class */ (function () {
    function NSheetDirective(sheetService, http, el) {
        this.sheetService = sheetService;
        this.http = http;
        this.el = el;
        this.onsuccess = new EventEmitter();
        this.onerror = new EventEmitter();
    }
    /**
     * @return {?}
     */
    NSheetDirective.prototype.ngOnInit = /**
     * @return {?}
     */
    function () {
        var _this = this;
        this.getSheet()
            .then((/**
         * @param {?} success
         * @return {?}
         */
        function (success) { return _this.onsuccess.emit(success); }))
            .catch((/**
         * @param {?} error
         * @return {?}
         */
        function (error) { return _this.onerror.emit(error); }));
    };
    /**
     * @return {?}
     */
    NSheetDirective.prototype.getSheet = /**
     * @return {?}
     */
    function () {
        var _this = this;
        return new Promise((/**
         * @param {?} resolve
         * @param {?} reject
         * @return {?}
         */
        function (resolve, reject) {
            _this.sheetService.getSheet(_this.sheetOptions, _this.htmlContent).then((/**
             * @param {?} res
             * @return {?}
             */
            function (res) {
                if (res.hasOwnProperty('excelAsHtml') && res['excelAsHtml'].length > 0 && !Array.isArray(res['excelAsHtml'])) {
                    if (!(typeof _this.sheetOptions === 'object' && _this.sheetOptions.hasOwnProperty('doAppend') && _this.sheetOptions.doAppend === true)) {
                        _this.el.nativeElement.innerHTML = "";
                    }
                    /** @type {?} */
                    var fragment = document.createRange().createContextualFragment(res['excelAsHtml']);
                    _this.el.nativeElement.appendChild(fragment);
                    return resolve(res);
                }
                else {
                    return reject(res);
                }
            })).catch((/**
             * @param {?} err
             * @return {?}
             */
            function (err) { return reject(err); }));
        }));
    };
    NSheetDirective.decorators = [
        { type: Directive, args: [{
                    selector: '[n-sheet]'
                },] }
    ];
    /** @nocollapse */
    NSheetDirective.ctorParameters = function () { return [
        { type: NSheetService },
        { type: HttpClient },
        { type: ElementRef }
    ]; };
    NSheetDirective.propDecorators = {
        sheetOptions: [{ type: Input }],
        htmlContent: [{ type: Input }],
        onsuccess: [{ type: Output }],
        onerror: [{ type: Output }]
    };
    return NSheetDirective;
}());
export { NSheetDirective };
if (false) {
    /** @type {?} */
    NSheetDirective.prototype.sheetOptions;
    /** @type {?} */
    NSheetDirective.prototype.htmlContent;
    /** @type {?} */
    NSheetDirective.prototype.onsuccess;
    /** @type {?} */
    NSheetDirective.prototype.onerror;
    /**
     * @type {?}
     * @private
     */
    NSheetDirective.prototype.sheetService;
    /**
     * @type {?}
     * @private
     */
    NSheetDirective.prototype.http;
    /**
     * @type {?}
     * @private
     */
    NSheetDirective.prototype.el;
}
