// neutrinos-seed-services/fesm2015/neutrinos-seed-services.js
import { os, name, version } from 'platform';
import { Injectable, EventEmitter, Output, Injector, Component, Input, NgModule, APP_INITIALIZER } from '@angular/core';
import { NgForage, NgForageCache, NgForageConfig, NgForageModule } from 'ngforage';
import { Subject } from 'rxjs/Subject';
import { messaging } from 'firebase';
import { HttpClient, HttpHeaders, HttpErrorResponse, HTTP_INTERCEPTORS, HttpClientModule } from '@angular/common/http';
import 'rxjs/add/operator/map';
import { Router, RouterModule } from '@angular/router';
import { Observable } from 'rxjs/Observable';
import { Observable as Observable$1, BehaviorSubject } from 'rxjs/Rx';
import { CommonModule } from '@angular/common';
import { HttpModule } from '@angular/http';

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
class NSystemService {
    /**
     * @return {?}
     */
    static getInstance() {
        if (!this.instance) {
            this.instance = new NSystemService();
        }
        return this.instance;
    }
    constructor() {
        this._deviceType = this.checkDevice();
    }
    /**
     * Returna 'mobile' or 'browser' based on the platform
     * @return {?}
     */
    checkDevice() {
        if (window['device'] && (window['device']['platform'] != 'browser' || window['device']['platform'] != 'Browser')) {
            return 'mobile';
        }
        else if (window['device'] && (window['device']['platform'] == 'browser' || window['device']['platform'] == 'Browser')) {
            return 'cordova_browser';
        }
        else {
            return 'browser';
        }
    }
    /**
     * @return {?}
     */
    get deviceType() {
        return this._deviceType;
    }
    /**
     * @param {?} key
     * @return {?}
     */
    getVal(key) {
        if (!NSystemService.envProperties) {
            NSystemService.envProperties = window['neutrinos']['environments'];
            this.properties = NSystemService.envProperties.properties;
        }
        if (key == 'properties') {
            if (NSystemService.envProperties) {
                return NSystemService.envProperties.properties;
            }
            else {
                return undefined;
            }
        }
        else {
            if (NSystemService.envProperties && NSystemService.envProperties.properties) {
                return NSystemService.envProperties.properties[key];
            }
            else {
                return undefined;
            }
        }
    }
    /**
     * @return {?}
     */
    isAndroid() {
        if (window['device'] && window['device']['platform'] == 'Android') {
            return true;
        }
        else {
            return false;
        }
    }
    /**
     * @return {?}
     */
    isIOS() {
        if (window['device'] && window['device']['platform'] == 'iOS') {
            return true;
        }
        else {
            return false;
        }
    }
    /**
     * @return {?}
     */
    getAndroidVersion() {
        if (this.isAndroid()) {
            return window['device']['version'];
        }
        else {
            return '';
        }
    }
    /**
     * @return {?}
     */
    getAuthUrl() {
        // this.properties = this.getVal('properties');
        return this.properties.baseUrl + this.properties.tenantName + '/auth/';
    }
    /**
     * @return {?}
     */
    getTenantUrl() {
        return this.properties.baseUrl + this.properties.tenantName + '/';
    }
    /**
     * @return {?}
     */
    getAppUrl() {
        return this.properties.baseUrl + this.properties.tenantName + '/proxy/' + this.properties.appName + '/';
    }
    /**
     * @return {?}
     */
    getProxyUrl() {
        return this.properties.baseUrl + this.properties.tenantName + '/proxy/';
    }
    /**
     * @return {?}
     */
    getDataModelUrl() {
        return this.properties.baseUrl + this.properties.tenantName + '/datamodel/' + this.properties.appDataSource + '/' + this.properties.appName + '/';
    }
    /**
     * @return {?}
     */
    getFileIOUrl() {
        return this.properties.baseUrl + this.properties.tenantName + '/fileio/' +
            this.properties.appDataSource + '/' + this.properties.appName + '/';
    }
    /**
     * @param {?} type
     * @return {?}
     */
    getPlatformDetails(type) {
        const /** @type {?} */ pDetails = {
            platform: os.family,
            platformVersion: os.version,
        };
        if (type === 'browser') {
            return Object.assign(pDetails, {
                type: 'browser',
                browser: name,
                browserVersion: version,
            });
        }
        else if (type === 'mobile' || type === 'cordova_browser') {
            return Object.assign(pDetails, {
                type: 'mobile',
                model: window['device'].model,
                manufacturer: window['device'].manufacturer,
                isVirtual: window['device'].isVirtual
            });
        }
    }
}

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
class NSessionStorageService {
    constructor() {
        NSessionStorageService.sessionStorageCache = sessionStorage;
    }
    /**
     * @return {?}
     */
    getSessionStorage() {
        return NSessionStorageService.sessionStorageCache;
    }
    /**
     * @param {?} key
     * @param {?} value
     * @return {?}
     */
    setValue(key, value) {
        NSessionStorageService.sessionStorageCache[key] = value;
        sessionStorage.setItem(key, value);
    }
    /**
     * @param {?} key
     * @return {?}
     */
    getValue(key) {
        if (!NSessionStorageService.sessionStorageCache[key]) {
            return null;
        }
        try {
            const /** @type {?} */ obj = NSessionStorageService.sessionStorageCache[key];
            return JSON.parse(obj);
        }
        catch (/** @type {?} */ error) {
            return NSessionStorageService.sessionStorageCache[key];
        }
    }
    /**
     * @param {?} key
     * @return {?}
     */
    remove(key) {
        if (NSessionStorageService.sessionStorageCache.hasOwnProperty(key)) {
            delete NSessionStorageService.sessionStorageCache[key];
            sessionStorage.removeItem(key);
        }
    }
    /**
     * @return {?}
     */
    clearSessionStorage() {
        NSessionStorageService.sessionStorageCache = {};
        sessionStorage.clear();
    }
}
NSessionStorageService.sessionStorageCache = {};
NSessionStorageService.decorators = [
    { type: Injectable },
];
/** @nocollapse */
NSessionStorageService.ctorParameters = () => [];

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
class NUtility {
    constructor() { }
    /**
     * @return {?}
     */
    generateUUID() {
        return this.__s4() + this.__s4() + '-' + this.__s4() + '-' + this.__s4() + '-' + this.__s4() + '-' + this.__s4() + this.__s4() + this.__s4();
    }
    /**
     * @return {?}
     */
    __s4() {
        return Math.floor((1 + Math.random()) * 0x10000)
            .toString(16)
            .substring(1);
    }
}

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
class NLocalStorageService {
    /**
     * @param {?=} ngfConfig
     * @param {?=} ngf
     * @param {?=} ngfCache
     */
    constructor(ngfConfig, ngf, ngfCache) {
        this.ngfConfig = ngfConfig;
        this.ngf = ngf;
        this.ngfCache = ngfCache;
        this.storageCache = {};
    }
    /**
     * @return {?}
     */
    initStorage() {
        return new Promise((resolve, reject) => {
            if (window['cordova']) {
                this.initNgForage();
            }
            this.ngf.iterate((value, key, iteratonNumber) => {
                this.storageCache[key] = value;
            }).then(result => {
                this.checkDeviceId();
                return resolve('iteration is completed');
            }).catch(error => {
                return reject(error);
            });
        });
    }
    /**
     * @return {?}
     */
    getStorage() {
        return this.storageCache;
    }
    /**
     * @param {?} key
     * @param {?} value
     * @return {?}
     */
    setValue(key, value) {
        if (window['cordova']) {
            this.initNgForage();
        }
        this.storageCache[key] = value;
        return this.ngf.setItem(key, value).then(result => {
            return result;
        }, error => {
            console.log(error);
        });
    }
    /**
     * @param {?} key
     * @return {?}
     */
    getValue(key) {
        if (!this.storageCache[key]) {
            return null;
        }
        try {
            const /** @type {?} */ obj = this.storageCache[key];
            return JSON.parse(obj);
        }
        catch (/** @type {?} */ error) {
            return this.storageCache[key];
        }
    }
    /**
     * @param {?} key
     * @return {?}
     */
    remove(key) {
        delete this.storageCache[key];
        if (window['cordova']) {
            this.initNgForage();
        }
        this.ngf.removeItem(key).then(fulfilled => {
            delete this.ngf[key];
        }).catch(error => {
            console.error('Could not remove', key);
        });
    }
    /**
     * @return {?}
     */
    clear() {
        this.storageCache = {};
        this.ngf.clear();
    }
    /**
     * @return {?}
     */
    pluginCheck() {
        if (window['cordova'] && window['NativeStorage']) {
            this.nativeStorageI = window['NativeStorage'];
            // return true;
        }
        // this.initStorage();
    }
    /**
     * @param {?} key
     * @return {?}
     */
    getItemNs(key) {
        return new Promise((resolve, reject) => {
            if (window['cordova'] && window['NativeStorage']) {
                this.nativeStorageI.getItem(key, result => {
                    resolve(result);
                }, error => {
                    reject(error);
                });
            }
        });
    }
    /**
     * @param {?} key
     * @param {?} value
     * @return {?}
     */
    setItemNs(key, value) {
        return new Promise((resolve, reject) => {
            if (window['cordova'] && window['NativeStorage']) {
                this.nativeStorageI.setItem(key, value, result => {
                    resolve(result);
                }, error => {
                    reject(error);
                });
            }
        });
    }
    /**
     * @param {?} key
     * @return {?}
     */
    removeItemNs(key) {
        return new Promise((resolve, reject) => {
            if (window['cordova'] && window['NativeStorage']) {
                this.nativeStorageI.remove(key, (result) => {
                    resolve(result);
                }, (error) => {
                    reject(error);
                });
            }
        });
    }
    /**
     * @return {?}
     */
    clearNs() {
        return new Promise((resolve, reject) => {
            if (window['cordova'] && window['NativeStorage']) {
                this.nativeStorageI.clear(result => {
                    resolve(result);
                }, error => {
                    reject(error);
                });
            }
        });
    }
    /**
     * @return {?}
     */
    initNgForage() {
        this.ngfConfig.configure({
            name: 'MyApp',
            driver: [
                NgForageConfig.DRIVER_WEBSQL,
            ]
        });
    }
    /**
     * @param {?} promise
     * @return {?}
     */
    promiseReflect(promise) {
        return promise.then(resolved => { return { v: resolved, status: 'resolved' }; }, error => { return { e: error, status: 'rejected' }; });
    }
    /**
     * @return {?}
     */
    clearLocalStorage() {
        this.remove('userObj');
        this.remove('accessToken');
        this.remove('refreshToken');
        this.remove('registrationId');
    }
    /**
     * Due to timing issues and circular dependency checkDeviceId is moved from NSystemService
     * @return {?}
     */
    checkDeviceId() {
        if (NSystemService.getInstance().checkDevice() === 'browser') {
            this._deviceUUID = this.getValue('uuid');
            if (!this._deviceUUID) {
                this._deviceUUID = new NUtility().generateUUID();
                this.setValue('uuid', this._deviceUUID);
            }
        }
        else {
            window['plugins'].uniqueDeviceID.get((uuid) => {
                this._deviceUUID = uuid;
                this.setValue('uuid', this._deviceUUID);
            });
        }
        return this._deviceUUID;
    }
    /**
     * @return {?}
     */
    get deviceUUID() {
        return this._deviceUUID;
    }
}
NLocalStorageService.decorators = [
    { type: Injectable },
];
/** @nocollapse */
NLocalStorageService.ctorParameters = () => [
    { type: NgForageConfig, },
    { type: NgForage, },
    { type: NgForageCache, },
];

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
class NTokenService {
    /**
     * @param {?} nLocalStorageService
     */
    constructor(nLocalStorageService) {
        this.nLocalStorageService = nLocalStorageService;
        this.nSessionStorage = new NSessionStorageService();
        // this.nLocalStorageService = new NLocalStorageService();
    }
    /**
     * @param {?} tokensObj
     * @param {?=} isRemember
     * @return {?}
     */
    updateTokens(tokensObj, isRemember) {
        let /** @type {?} */ userObj = this.decodeToken(tokensObj['accessToken']);
        this.nSessionStorage.setValue('accessToken', tokensObj['accessToken']);
        this.nSessionStorage.setValue('refreshToken', tokensObj['refreshToken']);
        this.nSessionStorage.setValue('userObj', JSON.stringify(userObj));
        if ((isRemember || this.nLocalStorageService.getValue('accessToken'))) {
            this.nLocalStorageService.setValue('accessToken', tokensObj['accessToken']);
            this.nLocalStorageService.setValue('refreshToken', tokensObj['refreshToken']);
            this.nLocalStorageService.setValue('userObj', JSON.stringify(userObj));
        }
    }
    /**
     * @return {?}
     */
    updateSessionStorage() {
        this.updateTokens(this.nLocalStorageService.getStorage(), true);
    }
    /**
     * @param {?} str
     * @return {?}
     */
    urlBase64Decode(str) {
        let /** @type {?} */ output = str.replace(/-/g, '+').replace(/_/g, '/');
        switch (output.length % 4) {
            case 0: {
                break;
            }
            case 2: {
                output += '==';
                break;
            }
            case 3: {
                output += '=';
                break;
            }
            default: {
                throw 'Illegal base64url string!';
            }
        }
        return this.b64DecodeUnicode(output);
    }
    /**
     * @param {?} str
     * @return {?}
     */
    b64decode(str) {
        let /** @type {?} */ chars = 'ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789+/=';
        let /** @type {?} */ output = '';
        str = String(str).replace(/=+$/, '');
        if (str.length % 4 == 1) {
            throw new Error("'atob' failed: The string to be decoded is not correctly encoded.");
        }
        for (
        // initialize result and counters
        let /** @type {?} */ bc = 0, /** @type {?} */ bs, /** @type {?} */ buffer, /** @type {?} */ idx = 0; 
        // get next character
        buffer = str.charAt(idx++); 
        // character found in table? initialize bit storage and add its ascii value;
        ~buffer && (bs = bc % 4 ? bs * 64 + buffer : buffer,
            // and if not first of each 4 characters,
            // convert the first 8 bits to one ascii character
            bc++ % 4) ? output += String.fromCharCode(255 & bs >> (-2 * bc & 6)) : 0) {
            // try to find character in table (0-63, not found => -1)
            buffer = chars.indexOf(buffer);
        }
        return output;
    }
    /**
     * @param {?} str
     * @return {?}
     */
    b64DecodeUnicode(str) {
        return decodeURIComponent(Array.prototype.map.call(this.b64decode(str), (c) => {
            return '%' + ('00' + c.charCodeAt(0).toString(16)).slice(-2);
        }).join(''));
    }
    /**
     * @param {?} token
     * @return {?}
     */
    decodeToken(token) {
        let /** @type {?} */ parts = token.split('.');
        if (parts.length !== 3) {
            throw new Error('JWT must have 3 parts');
        }
        let /** @type {?} */ decoded = this.urlBase64Decode(parts[1]);
        if (!decoded) {
            throw new Error('Cannot decode the token');
        }
        return JSON.parse(decoded);
    }
    /**
     * @param {?} token
     * @return {?}
     */
    getTokenExpirationDate(token) {
        let /** @type {?} */ decoded;
        decoded = this.decodeToken(token);
        if (!decoded.hasOwnProperty('exp')) {
            return null;
        }
        let /** @type {?} */ date = new Date(0); // The 0 here is the key, which sets the date to the epoch
        date.setUTCSeconds(decoded.exp);
        return date;
    }
    /**
     * @param {?} token
     * @param {?=} offsetSeconds
     * @return {?}
     */
    isTokenExpired(token, offsetSeconds) {
        let /** @type {?} */ date = this.getTokenExpirationDate(token);
        offsetSeconds = offsetSeconds || 0;
        if (date == null) {
            return false;
        }
        // Token expired?
        return !(date.valueOf() > (new Date().valueOf() + (offsetSeconds * 1000)));
    }
}
NTokenService.decorators = [
    { type: Injectable },
];
/** @nocollapse */
NTokenService.ctorParameters = () => [
    { type: NLocalStorageService, },
];

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
const /** @type {?} */ ServiceName = "PubSub Service";
class NPubSubService {
    constructor() {
        this.events = {};
    }
    /**
     * @param {?} event
     * @param {?=} callback
     * @param {?=} error
     * @param {?=} complete
     * @return {?}
     */
    $sub(event, callback, error, complete) {
        if (!event) {
            throw new Error(`[${ServiceName}] => Subscription method must get event name.`);
        }
        if (this.events[event] === undefined) {
            this.events[event] = new Subject();
        }
        if (typeof callback !== 'function') {
            return this.events[event].asObservable();
        }
        else {
            return this.events[event].asObservable().subscribe(callback, error, complete);
        }
    }
    /**
     * @param {?} event
     * @param {?=} eventObject
     * @return {?}
     */
    $pub(event, eventObject) {
        if (!event) {
            throw new Error(`[${ServiceName}] => Publish method must get event name.`);
        }
        else if (!this.events[event]) {
            return;
        }
        this.events[event].next(eventObject);
    }
}
NPubSubService.decorators = [
    { type: Injectable },
];
/** @nocollapse */
NPubSubService.ctorParameters = () => [];

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
class NLogoutService {
    /**
     * @param {?} nLocalStorageService
     */
    constructor(nLocalStorageService) {
        this.nLocalStorageService = nLocalStorageService;
        this.nSessionStorage = new NSessionStorageService();
    }
    /**
     * @return {?}
     */
    logout() {
        this.nSessionStorage.clearSessionStorage();
        this.nLocalStorageService.clearLocalStorage();
        return true;
    }
}
NLogoutService.decorators = [
    { type: Injectable },
];
/** @nocollapse */
NLogoutService.ctorParameters = () => [
    { type: NLocalStorageService, },
];

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
class NHTTPLoaderService {
    /**
     * @param {?} nLogoutService
     * @param {?} nPubSubService
     */
    constructor(nLogoutService, nPubSubService) {
        this.nLogoutService = nLogoutService;
        this.nPubSubService = nPubSubService;
        this.pendingRequests = 0;
        this.showLoading = false;
        this._isHTTPRequestInProgress = new Subject();
        this._isHTTPRequestInProgress$ = this._isHTTPRequestInProgress.asObservable();
    }
    /**
     * @param {?} bool
     * @return {?}
     */
    isHTTPRequestInProgress(bool) {
        if (bool) {
            this.pendingRequests++;
            if (!this.showLoading) {
                this._isHTTPRequestInProgress.next(bool);
            }
            this.showLoading = true;
        }
        else {
            this.pendingRequests--;
            if (this.pendingRequests <= 0) {
                if (this.showLoading) {
                    this._isHTTPRequestInProgress.next(bool);
                }
                this.showLoading = false;
            }
        }
    }
    /**
     * @param {?} error
     * @return {?}
     */
    alertError(error) {
        if (error.status < 200 || error.status > 500) {
            if (error.status === 0) {
                this.nPubSubService.$pub('openSnackBar', 'Connectivity issue');
            }
            else {
                this.nPubSubService.$pub('openSnackBar', 'Response failure');
            }
        }
        else if (error.status === 401) {
            this.nLogoutService.logout();
            if (error.error) {
                this.nPubSubService.$pub('openSnackBar', error.error);
            }
            else {
                this.nPubSubService.$pub('openSnackBar', error);
            }
        }
        else if (error.error) {
            if (error.error instanceof Object) {
                this.nPubSubService.$pub('openSnackBar', error.error.message);
            }
            else {
                this.nPubSubService.$pub('openSnackBar', error.error);
            }
        }
        else {
            this.nPubSubService.$pub('openSnackBar', 'Response failure');
        }
    }
}
NHTTPLoaderService.decorators = [
    { type: Injectable },
];
/** @nocollapse */
NHTTPLoaderService.ctorParameters = () => [
    { type: NLogoutService, },
    { type: NPubSubService, },
];

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
class NNotificationService {
    /**
     * @param {?} localStorageService
     * @param {?} pubSubService
     * @param {?} http
     * @param {?} bHttpLoader
     */
    constructor(localStorageService, pubSubService, http, bHttpLoader) {
        this.localStorageService = localStorageService;
        this.pubSubService = pubSubService;
        this.http = http;
        this.bHttpLoader = bHttpLoader;
        this.systemService = NSystemService.getInstance();
        this.possiblePushTypes = ['APNS', 'FCM'];
        this.firebaseSenderId = this.systemService.getVal('firebaseSenderId');
        this.isNotificationEnabled = this.systemService.getVal('isNotificationEnabled');
        this.appName = this.systemService.getVal('appName');
        this.deviceType = this.systemService.deviceType;
        this.sessionStorage = new NSessionStorageService();
        this.loginSubscribe = this.pubSubService.$sub('firebaseRegister', () => {
            this.enableNotification();
        });
    }
    /**
     * @return {?}
     */
    ngOnInit() {
    }
    /**
     * @return {?}
     */
    enableNotification() {
        let /** @type {?} */ pushType = this.getPushType(this.systemService.getVal('pushType'));
        document.addEventListener('deviceready', event => {
            if (this.isNotificationEnabled) {
                if (this.deviceType && this.deviceType != 'browser') {
                    this.deviceType = this.systemService.deviceType;
                    this.checkPermission(pushType).then(res => {
                        if (res) {
                            this.initializeNotifications(pushType);
                        }
                    });
                }
            }
        });
        if (this.isNotificationEnabled && pushType !== 'APNS') {
            if (this.deviceType && this.deviceType == 'browser' && window['Notification']) {
                this.initialiseWebPush();
            }
        }
    }
    /**
     * @return {?}
     */
    initialiseWebPush() {
        const /** @type {?} */ __this = this;
        const /** @type {?} */ messaging$$1 = messaging();
        messaging$$1.requestPermission()
            .then(function () {
            return messaging$$1.getToken();
        })
            .then(function (token) {
            if (token) {
                __this.sendRegDetails(token);
            }
        })
            .catch(function (err) {
            __this.bHttpLoader.alertError(err);
        });
        messaging$$1.onMessage(function (payload) {
            if (payload['notification']) {
                let /** @type {?} */ notificationObj = payload['notification'];
                let /** @type {?} */ options = {
                    body: notificationObj.body,
                    icon: notificationObj.icon
                };
                // creating a native browser message
                let /** @type {?} */ notificationUI = new Notification(notificationObj.title, options);
                notificationUI.onclick = function () {
                    window.focus(); // window is focused when the user clicks the notification using this
                };
            }
        });
    }
    /**
     * @param {?=} pushType
     * @return {?}
     */
    checkPermission(pushType) {
        // Android & iOS only
        // Checks whether the push notification permission has been granted.
        return new Promise((resolve) => {
            pushType = this.getPushType(pushType);
            if ((this.deviceType === 'Android' || this.deviceType === 'iOS') && (pushType === 'FCM')) {
                PushNotification.hasPermission(function (data) {
                    return resolve(data.isEnabled);
                });
            }
            else if (this.deviceType === 'iOS' && pushType === 'APNS') {
                APNSPushNotification.hasPermission(function (data) {
                    return resolve(data.isEnabled);
                });
            }
            else {
                return resolve(true);
            }
        });
    }
    /**
     * @param {?=} pushType
     * @return {?}
     */
    initializeNotifications(pushType) {
        //pushType = pushType ? pushType : 'FCM';
        pushType = this.getPushType(pushType);
        let /** @type {?} */ push;
        // Default if for FCM
        if (pushType === 'FCM') {
            push = window['PushNotification'].init({
                android: {
                    senderID: this.firebaseSenderId
                },
                ios: {
                    alert: "true",
                    badge: "true",
                    sound: "true",
                    senderID: this.firebaseSenderId
                },
            });
        }
        else if (pushType === 'APNS') {
            push = window['APNSPushNotification'].init({
                ios: {
                    alert: "true",
                    badge: "true",
                    sound: "true"
                }
            });
        }
        push.on('registration', (data) => {
            // data.registrationId
            this.sendRegDetails(data.registrationId);
        });
        // ToDo Christy get call back function from app user to change what happens once a notification arrives
        push.on('notification', (data) => {
            window['cordova'].plugins.notification.local.schedule({
                title: data.title,
                text: data.message,
                sound: data.sound,
                at: new Date().getTime()
            });
        });
        push.on('error', (e) => {
            // e.message
            console.error(e);
        });
    }
    /**
     * @param {?} registrationId
     * @return {?}
     */
    sendRegDetails(registrationId) {
        this.localStorageService.setValue('registrationId', registrationId);
        var /** @type {?} */ url = this.systemService.getTenantUrl() + 'notification/' + this.systemService.getVal('appName') + '/register';
        let /** @type {?} */ pushType = this.getPushType(this.systemService.getVal('pushType'));
        this.http.post(url, {
            'key': this.sessionStorage.getValue('userObj')['userKey'],
            'uuid': this.localStorageService.getValue('uuid'),
            'fbregid': registrationId,
            'pushType': pushType
        }).subscribe(result => {
            // this.pubSubService.$pub('FBRegComp');
        }, error => {
            console.log(error);
        });
    }
    /**
     * @param {?} currPushType
     * @return {?}
     */
    getPushType(currPushType) {
        let /** @type {?} */ isValidPush = typeof currPushType !== 'undefined' && this.possiblePushTypes.includes(currPushType.toUpperCase());
        let /** @type {?} */ pushType = isValidPush ? currPushType.toUpperCase() : 'FCM';
        return pushType;
    }
    /**
     * @return {?}
     */
    ngOnDestroy() {
        this.loginSubscribe.unSubscribe();
    }
}
NNotificationService.decorators = [
    { type: Injectable },
];
/** @nocollapse */
NNotificationService.ctorParameters = () => [
    { type: NLocalStorageService, },
    { type: NPubSubService, },
    { type: HttpClient, },
    { type: NHTTPLoaderService, },
];

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
class NLoginService {
    /**
     * @param {?} http
     * @param {?} pubSubService
     * @param {?} notificationService
     * @param {?} nLocalStorageService
     * @param {?} nTokenService
     */
    constructor(http, pubSubService, notificationService, nLocalStorageService, nTokenService) {
        this.http = http;
        this.pubSubService = pubSubService;
        this.notificationService = notificationService;
        this.nLocalStorageService = nLocalStorageService;
        this.nTokenService = nTokenService;
        this.loginCompleted = new EventEmitter();
        this.systemService = NSystemService.getInstance();
        // this.nTokenService = new NTokenService();
        this.nSessionStorage = new NSessionStorageService();
        // this.nLocalStorageService = new NLocalStorageService();
    }
    /**
     * @param {?} userName
     * @param {?} password
     * @param {?=} isRemember
     * @return {?}
     */
    login(userName, password, isRemember) {
        this.appProperties = this.systemService.getVal('properties');
        this.loginUrl = this.systemService.getAuthUrl() + this.appProperties.appName;
        this.uuid = this.nLocalStorageService.getValue('uuid');
        if (!this.uuid) {
            this.uuid = this.nLocalStorageService.checkDeviceId();
        }
        this.details = {
            username: userName,
            password: password,
        };
        this.details.platformDetails = this.systemService.getPlatformDetails(this.systemService.checkDevice());
        this.details.platformDetails['uuid'] = this.uuid;
        return this.http.post(this.loginUrl, JSON.stringify(this.details)).map(result => {
            const /** @type {?} */ tokensObj = result;
            if (tokensObj) {
                this.nTokenService.updateTokens(tokensObj, isRemember);
            }
            // TODO chris array of supported pushes currently only support APNS and Firebase
            if ((this.systemService.getVal('firebaseSenderId') != 'FIREBASE_SENDER_ID' && this.systemService.getVal('firebaseAuthKey') != 'FIREBASE_AUTH_KEY')
                || (this.systemService.getVal('pushType') === 'APNS' && this.systemService.isIOS())) {
                this.pubSubService.$pub('firebaseRegister');
            }
            this.pubSubService.$pub('loginComplete');
            return (result);
        }, error => {
            return (error);
        });
    }
    /**
     * @return {?}
     */
    isLoggedIn() {
        return this.nLocalStorageService.initStorage().then(result => {
            if (this.nSessionStorage.getValue('accessToken') && this.nSessionStorage.getValue('refreshToken') &&
                this.nSessionStorage.getValue('accessToken') != 'null' && this.nSessionStorage.getValue('refreshToken') != 'null') {
                return true;
            }
            return false;
        }).catch(error => {
            return false;
        });
    }
}
NLoginService.decorators = [
    { type: Injectable },
];
/** @nocollapse */
NLoginService.ctorParameters = () => [
    { type: HttpClient, },
    { type: NPubSubService, },
    { type: NNotificationService, },
    { type: NLocalStorageService, },
    { type: NTokenService, },
];
NLoginService.propDecorators = {
    "loginCompleted": [{ type: Output },],
};

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
class NAuthGuardService {
    /**
     * @param {?} nLoginService
     * @param {?} router
     */
    constructor(nLoginService, router) {
        this.nLoginService = nLoginService;
        this.router = router;
    }
    /**
     * @return {?}
     */
    canActivate() {
        return this.nLoginService.isLoggedIn().then(result => {
            if (result) {
                return true;
            }
            else {
                this.router.navigate(['unauthorized']);
                return false;
            }
        });
    }
}
NAuthGuardService.decorators = [
    { type: Injectable },
];
/** @nocollapse */
NAuthGuardService.ctorParameters = () => [
    { type: NLoginService, },
    { type: Router, },
];

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
class NDataModelService {
    /**
     * @param {?} http
     */
    constructor(http) {
        this.http = http;
        this.invalidDataModelName = 'Invalid data model name.';
        this.invalidDataModelId = 'Invalid data model id.';
        this.invalidDataModelObj = 'Invalid data model object.';
        this.systemService = NSystemService.getInstance();
    }
    /**
     *
     * @param {?} dataModelName
     * @param {?=} filter The filter query parameter allows to specify conditions on the documents to return.
     * The filter qparam value is any mongodb query… Defaults to {}
     * @param {?=} keys Projections to be applited on mongo db.
     * @param {?=} sort sort to be applied on the query results. Defaults to {}
     * @param {?=} pagenumber Page number for paginated queries. Defaults to 1
     * @param {?=} pagesize Size of each page to be returned. Defaults to 100.
     * @return {?}
     */
    get(dataModelName, filter, keys, sort, pagenumber, pagesize) {
        if (dataModelName) {
            // let modelNameUrl = `${this.getDataSourceURL(dataModelName)}${dataModelName}`;
            let /** @type {?} */ modelNameUrl = this.getDataSourceURL(dataModelName) + dataModelName;
            if (this.checkIfValid(filter) || this.checkIfValid(keys) || this.checkIfValid(sort) ||
                this.checkIfValid(pagenumber) || this.checkIfValid(pagesize)) {
                let /** @type {?} */ queryString = `${this.toQueryString({
                    'filter': filter,
                    'keys': keys,
                    'sort': sort,
                    'pagenumber': pagenumber,
                    'pagesize': pagesize
                })}`;
                if (queryString === '') {
                    queryString += '?filter={}';
                }
                else {
                    queryString = '?'.concat(queryString);
                }
                modelNameUrl += queryString;
            }
            return this.http.get(modelNameUrl).map((value, index) => {
                return value;
            }).catch(error => {
                return Observable.throw(error);
            });
        }
        else {
            return Observable.throw(new Error(`Could not get ${dataModelName}. ${this.invalidDataModelName}`));
        }
    }
    /**
     *
     * @param {?} dataModelName Data model name of the app
     * @param {?} dataModelObj Data Model object which is to be inserted
     * @return {?}
     */
    put(dataModelName, dataModelObj) {
        if (dataModelName) {
            if (dataModelObj) {
                // const modelNameUrl = `${this.getDataSourceURL(dataModelName)}${dataModelName}`;
                const /** @type {?} */ modelNameUrl = this.getDataSourceURL(dataModelName) + dataModelName;
                return this.http.put(modelNameUrl, dataModelObj).map((value, index) => {
                    return value;
                }).catch(error => {
                    return Observable.throw(error);
                });
            }
            else {
                return Observable.throw(new Error(`Could not put ${dataModelObj} in ${dataModelName}. ${this.invalidDataModelObj}`));
            }
        }
        else {
            return Observable.throw(new Error(`Could not put ${dataModelObj} in ${dataModelName}. ${this.invalidDataModelName}`));
        }
    }
    /**
     *
     * @param {?} dataModelName
     * @param {?} filter
     * @return {?}
     */
    delete(dataModelName, filter) {
        let /** @type {?} */ modelNameUrl;
        if (dataModelName) {
            // modelNameUrl = `${this.getDataSourceURL(dataModelName)}${dataModelName}`;
            modelNameUrl = this.getDataSourceURL(dataModelName) + dataModelName;
            if (this.checkIfValid(filter) && filter != '') {
                modelNameUrl += `?filter=${filter}`;
            }
            else {
                modelNameUrl += '?filter={}';
            }
            return this.http.delete(modelNameUrl).map((value, index) => {
                return value;
            }).catch(error => {
                return Observable.throw(error);
            });
        }
        else {
            return Observable.throw(new Error(`Could not delete ${dataModelName}. ${this.invalidDataModelName}`));
        }
    }
    /**
     *
     * @param {?} dataModelName Data model name which is to be updated
     * @param {?} updateObject
     * @return {?}
     */
    update(dataModelName, updateObject) {
        if (dataModelName && updateObject) {
            // const modelNameUrl = `${this.getDataSourceURL(dataModelName)}${dataModelName}`;
            const /** @type {?} */ modelNameUrl = this.getDataSourceURL(dataModelName) + dataModelName;
            return this.http.patch(modelNameUrl, updateObject).map((value, index) => {
                return value;
            }).catch(error => {
                return Observable.throw(error);
            });
        }
        else {
            return Observable.throw(new Error(`Could not update ${dataModelName}. ${this.invalidDataModelName}`));
        }
    }
    /**
     *
     * @param {?} dataModelName Data model name which is to be updated
     * @param {?} dataModelId Data model id which is to be updated
     * @return {?}
     */
    getById(dataModelName, dataModelId) {
        if (dataModelName) {
            if (dataModelId) {
                // const modelNameUrl = `${this.getDataSourceURL(dataModelName)}${dataModelName}/${dataModelId}`;
                const /** @type {?} */ modelNameUrl = this.getDataSourceURL(dataModelName) + dataModelName + "/" + dataModelId;
                return this.http.get(modelNameUrl).map((value, index) => {
                    return value;
                }).catch(error => {
                    return Observable.throw(error);
                });
            }
            else {
                Observable.throw(new Error(`Could not get ${dataModelName} by id ${dataModelId}. ${this.invalidDataModelId}`));
            }
        }
        else {
            Observable.throw(new Error(`Could not get ${dataModelName} by id ${dataModelId}. ${this.invalidDataModelName}`));
        }
    }
    /**
     *
     * @param {?} dataModelName Data model name which is to be deleted
     * @param {?} dataModelId Data model id which is to be deleted
     * @return {?}
     */
    deleteById(dataModelName, dataModelId) {
        if (dataModelName) {
            if (dataModelId) {
                // const modelNameUrl = `${this.getDataSourceURL(dataModelName)}${dataModelName}/${dataModelId}`;
                const /** @type {?} */ modelNameUrl = this.getDataSourceURL(dataModelName) + dataModelName + "/" + dataModelId;
                return this.http.delete(modelNameUrl).map((value, index) => {
                    return value;
                }).catch(error => {
                    return Observable.throw(error);
                });
            }
            else {
                Observable.throw(new Error(`Could not get ${dataModelName} by id ${dataModelId}. ${this.invalidDataModelId}`));
            }
        }
        else {
            return Observable.throw(new Error(`Could not delete ${dataModelName} by id ${dataModelId}. ${this.invalidDataModelName}`));
        }
    }
    /**
     *
     * @param {?} dataModelName Data model name which is to be update
     * @param {?} dataModelId Data model id which is to be updated
     * @param {?} dataModelObj Data Model object which is to be inserted
     * @return {?}
     */
    updateById(dataModelName, dataModelId, dataModelObj) {
        if (dataModelName) {
            if (dataModelId) {
                // const modelNameUrl = `${this.getDataSourceURL(dataModelName)}${dataModelName}/${dataModelId}`;
                const /** @type {?} */ modelNameUrl = this.getDataSourceURL(dataModelName) + dataModelName + "/" + dataModelId;
                var /** @type {?} */ dmObj = Object.assign({}, dataModelObj);
                delete dmObj['_id'];
                return this.http.patch(modelNameUrl, dmObj).map((value, index) => {
                    return value;
                }).catch(error => {
                    return Observable.throw(error);
                });
            }
            else {
                Observable.throw(new Error(`Could not get ${dataModelName} by id ${dataModelId}. ${this.invalidDataModelId}`));
            }
        }
        else {
            return Observable.throw(new Error(`Could not delete ${dataModelName} by id ${dataModelId}. ${this.invalidDataModelName}`));
        }
    }
    /**
     * @param {?} obj
     * @return {?}
     */
    toQueryString(obj) {
        const /** @type {?} */ parts = [];
        for (const /** @type {?} */ i in obj) {
            if (obj.hasOwnProperty(i) && this.checkIfValid(obj[i])) {
                parts.push((i) + '=' + JSON.stringify(obj[i]));
            }
        }
        return parts.join('&');
    }
    /**
     * @param {?} value
     * @return {?}
     */
    checkIfValid(value) {
        if (value === undefined || value == null) {
            return false;
        }
        else {
            return true;
        }
    }
    /**
     * @param {?} dataModelName
     * @return {?}
     */
    getDataSourceURL(dataModelName) {
        if (!this.dmDs) {
            this.dmDs = window['neutrinos']['dataSource'];
        }
        const /** @type {?} */ dsDm = this.dmDs[dataModelName];
        const /** @type {?} */ properties = this.systemService.properties;
        if (dsDm) {
            return properties.baseUrl + properties.tenantName + '/datamodel/' + dsDm + '/' + properties.appName + '/';
        }
        else {
            return this.systemService.getDataModelUrl();
        }
    }
}
NDataModelService.decorators = [
    { type: Injectable },
];
/** @nocollapse */
NDataModelService.ctorParameters = () => [
    { type: HttpClient, },
];

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
class NHttpService {
    /**
     * @param {?} nHTTPLoader
     * @param {?} inj
     * @param {?} nLocalStorageService
     * @param {?} nTokenService
     */
    constructor(nHTTPLoader, inj, nLocalStorageService, nTokenService) {
        this.nHTTPLoader = nHTTPLoader;
        this.inj = inj;
        this.nLocalStorageService = nLocalStorageService;
        this.nTokenService = nTokenService;
        this.timeout = 90000;
        this.isRefreshingToken = false;
        this.tokenSubject = new BehaviorSubject(null);
        this.systemService = NSystemService.getInstance();
        this.nSessionStorage = new NSessionStorageService();
        this.appProperties = this.systemService.getVal('properties');
        this.nPubSubService = new NPubSubService();
    }
    /**
     * @param {?} req
     * @param {?} next
     * @return {?}
     */
    intercept(req, next) {
        this.requestInterceptor();
        // Pass on the cloned request instead of the original request.
        return next.handle(this.requestOptions(req))
            .timeout(this.timeout)
            .catch(error => this.onCatch(error, req, next))
            .finally(() => {
            this.onFinally();
        });
    }
    ;
    /**
     * @param {?} error
     * @param {?} req
     * @param {?} next
     * @return {?}
     */
    updateToken(error, req, next) {
        if (this.appProperties.appAuthenticationStrategy === 'activeDirectory' ||
            this.appProperties.appAuthenticationStrategy === 'localAuth') {
            if (!this.isRefreshingToken) {
                this.isRefreshingToken = true;
                // Reset here so that the following requests wait until the token
                // comes back from the refreshToken call.
                this.tokenSubject.next(null);
                return this.refreshToken()
                    .switchMap((tokensObj) => {
                    if (tokensObj) {
                        this.nTokenService.updateTokens(tokensObj);
                        const /** @type {?} */ newToken = tokensObj['accessToken'];
                        this.tokenSubject.next(newToken);
                        return next.handle(this.requestOptions(req));
                    }
                    return Observable$1.throw(new Error('Can\'t refresh the token'));
                })
                    .catch(err => this.onCatchError(err))
                    .finally(() => this.isRefreshingToken = false);
            }
            else {
                return this.tokenSubject
                    .filter(token => token != null)
                    .take(1)
                    .switchMap(token => next.handle(this.requestOptions(req)));
            }
        }
        else {
            return this.onCatchError(error);
        }
    }
    /**
     * @return {?}
     */
    refreshToken() {
        const /** @type {?} */ http = this.inj.get(HttpClient);
        const /** @type {?} */ appProperties = this.systemService.getVal('properties');
        const /** @type {?} */ refreshUrl = this.systemService.getAuthUrl() + appProperties.appName + '/refresh';
        const /** @type {?} */ body = {
            'platformDetails': this.systemService.getPlatformDetails(this.systemService.checkDevice()),
            'userKey': this.nSessionStorage.getValue('userObj')['userKey'],
            'refreshToken': this.nSessionStorage.getValue('refreshToken')
        };
        body.platformDetails['uuid'] = this.nLocalStorageService.getValue('uuid');
        return http.post(refreshUrl, body);
    }
    /**
     * Request options.
     * @param {?=} req
     * @return {?} HttpRequest
     */
    requestOptions(req) {
        let /** @type {?} */ headers = req.headers;
        if (req.headers == null) {
            headers = new HttpHeaders();
        }
        req = req.clone({
            url: this.getFullUrl(req.url),
            headers: headers
        });
        const /** @type {?} */ baseUrl = NSystemService.getInstance().getVal('baseUrl');
        const /** @type {?} */ isArt = (baseUrl !== '' && req.url.includes(baseUrl));
        return isArt ? this.addDefaultHeaders(req) : req;
    }
    /**
     * Default options.
     * @param {?} req
     * @return {?} HttpHeadedrs
     */
    addDefaultHeaders(req) {
        /**
             * TODO: Add all default Headers over here
             */
        if (!req.headers.has('Access-Control-Allow-Origin')) {
            req.headers = req.headers.set('Access-Control-Allow-Origin', '*');
        }
        if (!req.headers.has('Content-Type')) {
            req.headers = req.headers.set('Content-Type', 'application/json');
        }
        else if (req.headers.has('Content-Type') && (req.headers.get('Content-Type') === 'no-content')) {
            req.headers = req.headers.delete('Content-Type');
        }
        if (!req.headers.has('Accept')) {
            req.headers = req.headers.set('Accept', 'application/json');
        }
        if (!req.headers.has('Authorization')) {
            this.appProperties = this.systemService.getVal('properties');
            if (this.appProperties && this.appProperties.appAuthenticationStrategy === 'basicAuth') {
                let /** @type {?} */ username, /** @type {?} */ password;
                if (this.appProperties.basicAuthUser && this.appProperties.basicAuthPassword) {
                    username = this.appProperties.basicAuthUser;
                    password = this.appProperties.basicAuthPassword;
                }
                else {
                    username = "bhive-art-proxyuser";
                    password = "password";
                    console.warn("Authentication strategy: Basic Auth. basicAuthUser and basicAuthPassword are not configured in environment. Setting default values.");
                }
                req.headers = req.headers.set('Authorization', 'Basic ' + btoa(username + ":" + password));
            }
            else if (this.appProperties && (this.appProperties.appAuthenticationStrategy === 'activeDirectory' ||
                this.appProperties.appAuthenticationStrategy === 'localAuth')) {
                if (this.nSessionStorage.getValue('accessToken')) {
                    req.headers = req.headers.set('Authorization', 'Bearer ' + this.nSessionStorage.getValue('accessToken'));
                }
            }
        }
        return req;
    }
    /**
     * Build API url.
     * @param {?} url
     * @return {?} string
     */
    getFullUrl(url) {
        // return full URL to API here
        return url;
    }
    /**
     * Request interceptor.
     * @return {?}
     */
    requestInterceptor() {
        this.nHTTPLoader.isHTTPRequestInProgress(true);
    }
    /**
     * Response interceptor.
     * @return {?}
     */
    responseInterceptor() {
        this.nHTTPLoader.isHTTPRequestInProgress(false);
    }
    /**
     * Error handler.
     * @param {?} error
     * @param {?} req
     * @param {?} next
     * @return {?} ErrorObservable
     */
    onCatch(error, req, next) {
        if (error instanceof HttpErrorResponse) {
            if ((/** @type {?} */ (error)).status === 403 && (/** @type {?} */ (error)).error.message === 'jwt expired') {
                return this.updateToken(error, req, next);
            }
            else {
                return this.onSubscribeError(error);
            }
        }
        else {
            return this.onSubscribeError(error);
        }
    }
    /**
     * onSubscribeError
     * @param {?} err
     * @return {?}
     */
    onSubscribeError(err) {
        this.nHTTPLoader.alertError(err);
        return this.onCatchError(err);
    }
    /**
     * onFinally
     * @return {?}
     */
    onFinally() {
        this.responseInterceptor();
    }
    /**
     * @param {?} error
     * @return {?}
     */
    onCatchError(error) {
        return Observable$1.throw(error);
    }
}
NHttpService.decorators = [
    { type: Injectable },
];
/** @nocollapse */
NHttpService.ctorParameters = () => [
    { type: NHTTPLoaderService, },
    { type: Injector, },
    { type: NLocalStorageService, },
    { type: NTokenService, },
];

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
class NSnackbarService {
    /**
     * @param {?} nPubSubService
     */
    constructor(nPubSubService) {
        this.nPubSubService = nPubSubService;
        this.snackService = new Subject();
        if (NSystemService.getInstance().getVal('useDefaultExceptionUI')) {
            this.snackBarListner = this.nPubSubService.$sub('openSnackBar', (msg) => {
                this.openSnackBar(msg);
            });
        }
    }
    /**
     * @return {?}
     */
    get() {
        return this.snackService.asObservable();
    }
    /**
     * @param {?} data
     * @return {?}
     */
    add(data) {
        this.snackService.next({
            action: 'add',
            data: data
        });
    }
    /**
     * @param {?} id
     * @return {?}
     */
    remove(id) {
        this.snackService.next({ action: 'remove', id: id });
    }
    /**
     * @return {?}
     */
    clear() {
        this.snackService.next({ action: 'clear' });
    }
    /**
     * @param {?} message
     * @param {?=} timeout
     * @return {?}
     */
    openSnackBar(message, timeout) {
        this.add({
            msg: message,
            timeout: timeout || 5000,
            action: {
                text: 'Close',
                onClick: (snack) => {
                },
            },
            onAdd: (snack) => {
            },
            onRemove: (snack) => {
            }
        });
    }
}
NSnackbarService.decorators = [
    { type: Injectable },
];
/** @nocollapse */
NSnackbarService.ctorParameters = () => [
    { type: NPubSubService, },
];

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
class SnackComponent {
}
SnackComponent.decorators = [
    { type: Component, args: [{
                selector: 'n-snack',
                template: `
     <div class="snack" [ngStyle]="{background: background ? background : '#323232' , color: color ? color : '#fff'}" [ngClass]=" ['snackbar-layout', 'customClass' ]">
          <ng-content></ng-content>
       </div> 
  `,
                styles: [`.snack-action{cursor:pointer;color:#2196f3;font-weight:700}.container{display:flex;flex-direction:row;flex-wrap:wrap}.child{width:60%;flex:1 0;margin-right:.5em}.snackbars.bottom-center{align-items:middle;left:50%;-webkit-transform:translate(-50%,0);transform:translate(-50%,0);bottom:1px;align-items:center}.snackbars{position:fixed;z-index:99999;max-width:100%;word-wrap:break-word;display:inline;font-family:Roboto;letter-spacing:1px}.snack{padding:1em;border-radius:.3em;max-height:50vh;overflow:auto}`]
            },] },
];
/** @nocollapse */
SnackComponent.propDecorators = {
    "background": [{ type: Input },],
    "color": [{ type: Input },],
    "customClass": [{ type: Input },],
};

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
class SnackbarComponent {
    /**
     * @param {?} snackbarService
     */
    constructor(snackbarService) {
        this.snackbarService = snackbarService;
        this.max = 1;
        this.onAdd = new EventEmitter();
        this.onRemove = new EventEmitter();
        this.onClear = new EventEmitter();
        this.snacks = [];
        this.snackbarService.get()
            .subscribe(snack => {
            if (snack.action === 'add') {
                this.add(snack.data);
            }
            else if (snack.action === 'remove') {
                this.remove(snack.id);
            }
            else if (snack.action === 'clear') {
                this.clear();
            }
        });
    }
    /**
     * @param {?} snack
     * @return {?}
     */
    add(snack) {
        let /** @type {?} */ timeout;
        const /** @type {?} */ id = this.uuid();
        if (this.max && this.max > 0 && this.snacks.length === this.max) {
            this.remove(this.snacks[0].id);
        }
        if (snack.timeout || this.timeout) {
            timeout = setTimeout(() => {
                this.remove(id);
            }, snack.timeout || this.timeout);
        }
        const /** @type {?} */ data = Object.assign({ id: id, timeoutObj: timeout }, snack);
        if (snack.action) {
            const /** @type {?} */ that = this;
            const /** @type {?} */ fcn = snack.action.onClick || new Function();
            snack.action.onClick = () => {
                fcn(data);
                that.remove(id);
            };
        }
        if (snack.onAdd) {
            snack.onAdd(data);
        }
        if (this.onAdd) {
            this.onAdd.emit(data);
        }
        this.snacks.push(data);
    }
    /**
     * @param {?} id
     * @return {?}
     */
    remove(id) {
        const /** @type {?} */ snack = this.snacks.find(obj => obj.id === id);
        if (snack) {
            if (snack.onRemove) {
                snack.onRemove(snack);
            }
            if (this.onRemove) {
                this.onRemove.emit(snack);
            }
            if (snack.timeoutObj) {
                clearTimeout(snack.timeoutObj);
            }
        }
        this.snacks = this.snacks.filter(obj => obj.id !== id);
    }
    /**
     * @return {?}
     */
    clear() {
        // this.snacks.forEach(snack => {
        //   this.remove(snack.id);
        // });
        this.snacks = [];
        if (this.onClear) {
            this.onClear.emit(true);
        }
    }
    /**
     * @return {?}
     */
    uuid() {
        // tslint:disable:no-bitwise
        return 'xxxxxxxx-xxxx-4xxx-yxxx-xxxxxxxxxxxx'.replace(/[xy]/g, function (c) {
            const /** @type {?} */ r = Math.random() * 16 | 0, /** @type {?} */ v = c === 'x' ? r : (r & 0x3 | 0x8);
            return v.toString(16);
        });
        // tslint:enable:no-bitwise
    }
    /**
     * @param {?} background
     * @return {?}
     */
    calcTextColor(background) {
        if (!background) {
            return null;
        }
        /**
         * @param {?} hex
         * @return {?}
         */
        function hexToRgb(hex) {
            const /** @type {?} */ shorthandRegex = /^#?([a-f\d])([a-f\d])([a-f\d])$/i;
            hex = hex.replace(shorthandRegex, (m, r, g, b) => {
                return r + r + g + g + b + b;
            });
            const /** @type {?} */ result = /^#?([a-f\d]{2})([a-f\d]{2})([a-f\d]{2})$/i.exec(hex);
            return result ? {
                r: parseInt(result[1], 16),
                g: parseInt(result[2], 16),
                b: parseInt(result[3], 16)
            } : null;
        }
        const /** @type {?} */ rgb = hexToRgb(background);
        if (!rgb) {
            return null;
        }
        const /** @type {?} */ color = [rgb.r / 255, rgb.g / 255, rgb.b / 255];
        for (let /** @type {?} */ i = 0; i < color.length; ++i) {
            if (color[i] <= 0.03928) {
                color[i] = color[i] / 12.92;
            }
            else {
                color[i] = Math.pow((color[i] + 0.055) / 1.055, 2.4);
            }
        }
        const /** @type {?} */ l = 0.2126 * color[0] + 0.7152 * color[1] + 0.0722 * color[2];
        if (l > 0.179) {
            return '#000';
        }
        else {
            return '#fff';
        }
    }
}
SnackbarComponent.decorators = [
    { type: Component, args: [{
                selector: 'n-snackbar',
                template: `
      <div class="snackbars" [ngClass]="position || 'bottom-center'">
          <n-snack   *ngFor="let snackbar of snacks" [background]="snackbar.background || background"
                        [customClass]="snackbar.customClass || customClass"
                        [color]="snackbar.color || color || calcTextColor(snackbar.background || background)">
              <div class="container">
              <div class="snack-text child" >
                  {{snackbar.msg}}
              </div>
              <div *ngIf="snackbar.action.text" class="snack-action" (click)="snackbar.action.onClick()"
                   [ngStyle]="{color: snackbar.action.color || accent}">
                  {{snackbar.action.text}}
              </div>
              </div>
          </n-snack>
      </div>
  `,
                styles: [`.snack-action{cursor:pointer;color:#2196f3;font-weight:700}.container{display:flex;flex-direction:row;flex-wrap:wrap}.child{width:60%;flex:1 0;margin-right:.5em}.snackbars.bottom-center{align-items:middle;left:50%;-webkit-transform:translate(-50%,0);transform:translate(-50%,0);bottom:1px;align-items:center}.snackbars{position:fixed;z-index:99999;max-width:100%;word-wrap:break-word;display:inline;font-family:Roboto;letter-spacing:1px}.snack{padding:1em;border-radius:.3em;max-height:50vh;overflow:auto}`]
            },] },
];
/** @nocollapse */
SnackbarComponent.ctorParameters = () => [
    { type: NSnackbarService, },
];
SnackbarComponent.propDecorators = {
    "position": [{ type: Input },],
    "max": [{ type: Input },],
    "background": [{ type: Input },],
    "accent": [{ type: Input },],
    "color": [{ type: Input },],
    "customClass": [{ type: Input },],
    "timeout": [{ type: Input },],
    "onAdd": [{ type: Output },],
    "onRemove": [{ type: Output },],
    "onClear": [{ type: Output },],
};

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
/**
 * @param {?} startupService
 * @return {?}
 */
function startupServiceFactory(startupService) {
    const /** @type {?} */ res = () => {
        // console.log('module before initstorage call');
        return startupService.initStorage();
    };
    return res;
}
class NeutrinosSeedServicesModule {
}
NeutrinosSeedServicesModule.decorators = [
    { type: NgModule, args: [{
                imports: [
                    CommonModule,
                    HttpClientModule,
                    RouterModule,
                    HttpModule,
                    NgForageModule.forRoot()
                ],
                providers: [
                    {
                        provide: HTTP_INTERCEPTORS,
                        useClass: NHttpService,
                        multi: true
                    },
                    {
                        // Provider for APP_INITIALIZER
                        provide: APP_INITIALIZER,
                        useFactory: startupServiceFactory,
                        deps: [NLocalStorageService],
                        multi: true
                    },
                    NAuthGuardService,
                    NDataModelService,
                    NHttpService,
                    NHTTPLoaderService,
                    NLocalStorageService,
                    NLoginService,
                    NLogoutService,
                    NNotificationService,
                    NSnackbarService,
                    NPubSubService,
                    NSessionStorageService,
                    NTokenService
                ],
                declarations: [
                    SnackComponent,
                    SnackbarComponent
                ],
                exports: [
                    SnackComponent,
                    SnackbarComponent
                ]
            },] },
];

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */

export { NAuthGuardService, NDataModelService, NHttpService, NHTTPLoaderService, NLocalStorageService, NLoginService, NLogoutService, NNotificationService, NSnackbarService, NSystemService, NTokenService, NUtility, NPubSubService, NSessionStorageService, SnackbarComponent, startupServiceFactory, NeutrinosSeedServicesModule, SnackComponent as ɵa };
